### PR TITLE
completion Working in progress

### PR DESCRIPTION
**Pull request by BNMetrics**
Some of the completion is done, such as build labels(however this requires the file to be complete), properties.
More tests should be added to ensure correct functionalities.

### CODE REVIEW

File: tools/build_langserver/langserver/analyzer.go [hunk 1]
```diff
@@ -241,87 +269,171 @@ func (a *Analyzer) BuildLabelFromString(ctx context.Context, rootPath string,
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the content for the BuildDef
+	labelfileContent, err := ReadFile(ctx, lsp.DocumentURI(path))
+	if err != nil {
+		return nil, err
+	}
+	buildDef.Content = strings.Join(labelfileContent[buildDef.Pos.Line:buildDef.EndPos.Line+1], "\n")
```
Comment (peterebden) [inline tools/build_langserver/langserver/analyzer.go]: you sure about this? are the Pos's zero-indexed or 1-indexed?

File: tools/build_langserver/langserver/completion.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		}
+	}
+
+	// Map the labels to a lsp.CompletionItem slice
+	var completionList []*lsp.CompletionItem
+	for _, label := range labels {
+		partial := strings.Replace(label, lineContent, "", 1)
+		TERange := getTERange(pos, partial)
+		detail := "BUILD Label: label"
+		label := label
```
Comment (peterebden) [inline tools/build_langserver/langserver/completion.go]: unnecessary

File: tools/build_langserver/langserver/completion.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+func getTERange(pos lsp.Position, partial string) lsp.Range {
+	return lsp.Range{
+		Start: lsp.Position{Line: pos.Line, Character: pos.Character - len(partial)},
+		End:   pos,
+	}
+}
+
+func isVisible(buildDef *BuildDef, currentPkg string) bool {
+	for _, i := range buildDef.Visibility {
+		if i == "PUBLIC" || strings.HasPrefix(i, currentPkg) {
```
Comment (peterebden) [inline tools/build_langserver/langserver/completion.go]: This isn't quite correct... you should be using the `Includes` function of `BuildLabel`

File: tools/build_langserver/langserver/completion.go (hunk 2 above)
Comment (BNMetrics) [reply tools/build_langserver/langserver/completion.go]: oops lol

File: tools/build_langserver/langserver/analyzer.go (hunk 1 above)
Comment (BNMetrics) [reply tools/build_langserver/langserver/analyzer.go]: They are zero indexed! :) lsp.Positions! :)

File: tools/build_langserver/langserver_main.go [hunk 4]
```diff
@@ -40,20 +42,38 @@ func main() {
 
 	handler := langserver.NewHandler()
 
-	serve(handler)
+	if err := serve(handler); err != nil {
+		fmt.Fprintln(os.Stderr, err)
```
Comment (peterebden) [inline tools/build_langserver/langserver_main.go]: You should use `log.Fatalf` for this

**Pull request merged**